### PR TITLE
Introduce Go modules support

### LINKED ISSUE #1488: Go modules support
Since 1.11 is introducing official module support, does Linkerd have plans to drop dep? 1.11 final should be dropping in next month~ or sooner. It looks like the majority pain points have been addressed (such as how do we handle docker image caching to speed up builds, etc).

I am more than happy to provide a PR to support this.

### PR DESCRIPTION
**Pull request by siggy**
The repo relied on `dep` for managing Go dependencies. Go 1.11 shipped
with Go modules support. Go 1.13 will be released in August 2019 with
module support enabled by default, deprecating GOPATH.

This change replaces `dep` with Go modules for dependency management.
All scripts, including Docker builds and ci, should work without any dev
environment changes.

To execute `go` commands directly during development, do one of the
following:
1. clone this repo outside of `GOPATH`; or
2. run `export GO111MODULE=on`

Summary of changes:
- Docker build scripts and ci set `-mod=readonly`, to ensure
  dependencies defined in `go.mod` are exactly what is used for the
  builds.
- Dependency updates to `go.mod` are accomplished by running
 `go build` and `go test` directly.
- `bin/go-run`, `bin/build-cli-bin`, and `bin/test-run` set
  `GO111MODULE=on`, permitting usage inside and outside of GOPATH.
- `gcr.io/linkerd-io/go-deps` tags hashed from `go.mod`.
- `bin/update-codegen.sh` still requires running from GOPATH,
  instructions added to BUILD.md.

Fixes #1488

Signed-off-by: Andrew Seigner <siggy@buoyant.io>

### CODE REVIEW

**Comment by siggy** [comment]
For vscode users: https://github.com/Microsoft/vscode-go/wiki/Go-modules-support-in-Visual-Studio-Code#updates-as-of-090

**Comment by paranoidaditya** [comment]
What do you think about not vendoring at all?

We are caching `$GOPATH/pkg/mod` in Travis and we already seem to have an intermediate step pulling in the deps in the Dockerfiles

**Comment by siggy** [comment]
> What do you think about not vendoring at all?

@paranoidaditya AFAIU, we're not vendoring? Go modules relies on `$GOPATH/pkg/mod` as the source of dependencies.

File: BUILD.md [hunk 1]
```diff
@@ -318,20 +326,33 @@ DOCKER_TRACE=1 bin/docker-build-proxy
 ### Updating protobuf dependencies
  If you make Protobuf changes, run:
  ```bash
-bin/dep ensure
```
Comment (scottcarol) [inline BUILD.md]: Do you also want to remove the `bin/dep ensure` instructions on line 265 in the Web section?

File: BUILD.md [hunk 2]
```diff
@@ -318,20 +326,33 @@ DOCKER_TRACE=1 bin/docker-build-proxy
...
 
-`bin/update-go-deps-shas` must be run when go dependencies change.
+`bin/update-go-deps-shas` must be run when Go dependencies change.
+
+### Updating ServiceProfile generated code
+
+The [ServiceProfile client code](./controller/gen/client) is generated by
+[`bin/update-codegen.sh`](bin/update-codegen.sh), which depends on
+[K8s code-generator](https://github.com/kubernetes/code-generator), which does
+not yet support Go Modules. To re-generate this code, check out this repo into
```
Comment (adleong) [inline BUILD.md]: For an example of doing this with Go Modules, see https://github.com/deislabs/smi-sdk-go/blob/master/hack/update-codegen.sh

Comment (siggy) [reply BUILD.md]: filed #3145 to track, thanks!

File: .travis.yml [hunk 3]
```diff
@@ -22,23 +22,22 @@ jobs:
     - language: go
       # Quote the version number to avoid parsing issues like
       # https://github.com/travis-ci/gimme/issues/132.
-      go: "1.11.5"
+      go: "1.12.7"
       go_import_path: github.com/linkerd/linkerd2
+      env:
+        - GO111MODULE=on
```
Comment (alpeb) [inline .travis.yml]: Is this because Travis forces us to have the code in `github.com/linkerd/linkerd2`?

Comment (siggy) [reply .travis.yml]: Yeah that's my understanding:
https://docs.travis-ci.com/user/languages/go/#go-import-path

Eventually we should be able to move away from always setting `GO111MODULE=on`, once we're confident all environments are checked out outside of the GOPATH.

File: bin/protoc-go.sh [hunk 4]
```diff
@@ -3,21 +3,31 @@
...
-
-    rm -rf controller/gen/common controller/gen/controller controller/gen/public controller/gen/config
-    mkdir -p controller/gen
-
     for f in $@; do
-        bin/protoc -I proto --go_out="plugins=grpc:$GOPATH/src" "$f"
+        bin/protoc -I proto --go_out=plugins=grpc,paths=source_relative:controller/gen "$f"
     done
-
-    git add controller/gen
```
Comment (alpeb) [inline bin/protoc-go.sh]: Nice. I also didn't like protoc messing with git :+1:

Comment (siggy) [reply bin/protoc-go.sh]: @alpeb then you're not going to like line 33 😢

Comment (alpeb) [reply bin/protoc-go.sh]: oh haha, I missed that.

**Pull request merged**